### PR TITLE
[docs] Add VLA full hidden_states export example

### LINKED ISSUE #4414: how to return hidden_states?
use gather_context_logits and gather_generation_logits can help to return logits. but how to return hidden_states? plz help, thx

### PR DESCRIPTION
**Pull request by hous-lab**
## Summary

Adds `examples/vla_hidden_states_export/` - a guide for exporting full 3D hidden_states from TensorRT-LLM for VLA (Vision-Language-Action) models.

Fixes #4414

## Background

VLA models (Orion, OpenVLA, RT-2) use the LLM's hidden states as input to downstream task heads, rather than for token generation:

- Standard LLM: input -> LLM -> logits -> tokens -> text
- VLA model: images+text -> LLM -> **hidden_states** -> planning head -> trajectory

The downstream head needs the hidden state at a specific token position (e.g. a "waypoint" token), which requires the complete `[batch, seq_len, hidden_dim]` tensor - not just the last token.

## Why existing APIs fall short

- **gather_last_token_logits**: compresses 3D to 2D, only last token survives
- **additional_model_outputs** (v1.1+): standard `DecoderModelForCausalLM.forward` only returns logits; hidden_states is consumed by LogitsProcessor before it can be exposed
- **SaveHiddenStatesDecodingConfig**: saves to disk for offline EAGLE3 training, not usable at inference

### Solution A: TRT backend mark_output (v0.7-v0.21, verified)

Insert `mark_output` before `gather_last_token_logits`:

```python
hidden_states.mark_output('full_hidden_states', self.config.dtype)
hidden_states = gather_last_token_logits(...)
lm_logits = self.lm_head(hidden_states)
```

Compatibility verified across v0.7.0 to v0.21.0 (15 versions).

Reading at inference:
```python
full_hs = model.session.debug_buffer["full_hidden_states"]
ego_feature = full_hs[0, waypoint_idx, :]
```

### Solution B: PyTorch backend forward modification (v1.x)

Attach the raw 3D tensor before LogitsProcessor compresses it, enabling `additional_model_outputs=["hidden_states"]`.

## Verification (Solution A)

Tested on Orion VLA model (ICCV'25) with TRT-LLM v0.13.0:

- Engine output: `[1, 599, 4096]` (full 3D)
- Hidden_states CosSim vs PyTorch: **0.9994** (INT8)
- End-to-end plan_L2_1s: **0.686** (PyTorch: 0.690)
- Platforms: RTX 4090 + AGX Orin

## Notes

- Solution B is based on v1.x source code analysis, not yet runtime-tested
- Happy to adjust based on maintainer feedback on preferred integration approach

## Dev Engineer Review

- Added documentation, patches, scripts, and inference code for exporting full 3D hidden states.
- Documented TRT-LLM v0.x support through `mark_output` before `gather_last_token_logits`.
- Documented a PyTorch v1.x approach through `additional_model_outputs`.
- Added CUDA inference code that validates engine I/O, retrieves `full_hidden_states`, and extracts token features.
- Added version guidance and Orion verification results.
- The PyTorch approach has not been runtime-tested.
- No public API declarations, configuration files, or test-list files were changed.
- Review focus: verify CUDA buffer handling, tensor-rank support, sequence-length assumptions, API compatibility, and error handling in `HiddenStatesEngine`.
- Verdict: needs follow-up because the new inference path and PyTorch integration are not fully runtime-validated.

## QA Engineer Review

- Added `test_engine_has_full_hidden_states`.
- Added `test_hidden_states_is_3d`.
- Added `test_token_extraction`.
- Added a command-line test entry point with simulated extraction fallback.
- These tests are not listed in `tests/integration/test_lists/` test-db or QA files.
- Engine tests require a suitable TensorRT-LLM engine and are not standard CI tests.
- Verdict: needs follow-up.

### CODE REVIEW

File: examples/vla_hidden_states_export/patches/modeling_utils_v1x.patch [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+             inputs_embeds=inputs_embeds,
+             spec_metadata=spec_metadata,
+             lora_params=lora_params,
+         )
++        # [VLA] Attach full 3D hidden_states before LogitsProcessor compresses it.
++        # This enables VLA models to access hidden_states at any token position
++        # for downstream task heads (e.g., planning, action prediction).
++        # Usage: SamplingParams(additional_model_outputs=["hidden_states"])
++        # See: https://github.com/NVIDIA/TensorRT-LLM/issues/4414
++        if hasattr(self, '_additional_output_names') and 'hidden_states' in self._additional_output_names:
```
Comment (brnguyen2) [inline examples/vla_hidden_states_export/patches/modeling_utils_v1x.patch]: This patch is a no-op as written: nothing in the codebase ever sets `_additional_output_names`, and nothing ever reads `self._additional_outputs`, so `additional_model_outputs=["hidden_states"]` would still come back empty.

The framework's actual contract is dict-return from `forward`. `model_engine.py:6611` checks `isinstance(outputs, dict)`, and `HandleAdditionalOutputs` (`py_executor.py:6572`) looks up each name in `SamplingParams.additional_model_outputs` as a key of that dict. So the working version is what your README snippet already shows:

```python
return {
    "logits": self.logits_processor.forward(output, self.lm_head, attn_metadata, return_context_logits),
    "hidden_states": output,
}
```

Since that path already exists, a model subclass overriding `forward` this way needs no core patch at all — which would make a much better example than a patch file.

File: examples/vla_hidden_states_export/patches/modeling_utils_v0x.patch [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+#         the engine, before gather_last_token_logits compresses it to 2D.
+#
+# Why: VLA models (Orion, OpenVLA) need the full hidden_states to feed
+#      downstream planning heads, not just the last token's logits.
+#
+# Related issue: https://github.com/NVIDIA/TensorRT-LLM/issues/4414
+
+--- a/tensorrt_llm/models/modeling_utils.py
++++ b/tensorrt_llm/models/modeling_utils.py
+@@ -XXX,7 +XXX,11 @@ class DecoderModelForCausalLM(DecoderModel):
```
Comment (brnguyen2) [inline examples/vla_hidden_states_export/patches/modeling_utils_v0x.patch]: `@@ -XXX,7 +XXX,11 @@` isn't a valid hunk header — `git apply` and `patch` will both reject this file. If a patch file is kept at all, generate it with `git diff` against a real checkout of the target version so the line numbers are concrete.

File: examples/vla_hidden_states_export/README.md [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+```
+
+### Solution B: PyTorch Backend — Forward Modification (v1.0+)
+
+For the PyTorch-native backend in v1.x:
+
+```python
+# tensorrt_llm/_torch/models/modeling_utils.py — DecoderModelForCausalLM.forward()
+def forward(self, ..., additional_model_outputs=None):
+    output = self.model(...)  # Full 3D [batch, seq_len, hidden_dim]
```
Comment (brnguyen2) [inline examples/vla_hidden_states_export/README.md]: `# Full 3D [batch, seq_len, hidden_dim]` is wrong for the PyTorch backend. With remove-input-padding (always on in the PyTorch path) the model returns packed tokens, shape `[num_tokens, hidden_dim]`, where `num_tokens` is the flattened sum over the batch. `HandleAdditionalOutputs` then slices per request using the context/generation token counts, so what a caller gets back is `[seq_len, hidden_dim]` for that request — no batch dimension. Please correct the shapes throughout the README; the "complete 3D tensor" framing in the intro sets up the wrong expectation.

File: examples/vla_hidden_states_export/inference_python.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        cuda.memcpy_htod(d_input, input_ids_np)
+        self.context.set_tensor_address("input_ids", int(d_input))
+        buffers["input_ids"] = d_input
+
+        # Output: full_hidden_states
+        hs_shape = tuple(self.context.get_tensor_shape("full_hidden_states"))
+        hs_size = int(np.prod([max(1, s) for s in hs_shape]))
+        hs_dtype = np.float16  # fp16 engine
+        d_hs = cuda.mem_alloc(hs_size * 2)  # fp16 = 2 bytes
+        self.context.set_tensor_shape("full_hidden_states", hs_shape)
```
Comment (brnguyen2) [inline examples/vla_hidden_states_export/inference_python.py]: Three problems make this unrunnable against a real engine:

1. `set_tensor_shape` is only valid for *input* tensors; calling it on `full_hidden_states` (an output) is an error.
2. `get_tensor_shape` is queried before any input shape is bound, so dynamic dims come back as `-1`. `max(1, s)` then turns those into 1, so `hs_size` is far smaller than the real output and `cuda.mem_alloc(hs_size * 2)` under-allocates — the engine writes past the buffer.
3. Only `input_ids` is bound; a TRT-LLM engine needs position ids, sequence lengths, KV-cache pointers etc., so `execute_async_v3` will fail on unbound tensors.

Also `pycuda` is not a TensorRT-LLM dependency. Rather than repairing this, the PyTorch-backend `llm.generate` + `additional_generation_outputs` path gives a working ~20-line example with no manual buffer management.

File: examples/vla_hidden_states_export/tests/test_hidden_states.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    python -m pytest tests/test_hidden_states.py -v
+    # or standalone:
+    python tests/test_hidden_states.py --engine_path /path/to/rank0.engine
+"""
+import argparse
+
+import numpy as np
+
+
+def test_engine_has_full_hidden_states(engine_path):
```
Comment (brnguyen2) [inline examples/vla_hidden_states_export/tests/test_hidden_states.py]: The module docstring documents `python -m pytest tests/test_hidden_states.py -v`, but `test_engine_has_full_hidden_states(engine_path)` and `test_hidden_states_is_3d(engine_path)` take a parameter that is not a fixture — pytest will error both with `fixture 'engine_path' not found`. Either add a `--engine-path` option plus an `engine_path` fixture in a `conftest.py`, or rename these to non-`test_` helpers and drop the pytest instruction.

File: examples/vla_hidden_states_export/tests/test_hidden_states.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+    """Verify we can extract a specific token's hidden state from the full tensor."""
+    # Simulated full_hidden_states [1, 599, 4096]
+    full_hs = np.random.randn(1, 599, 4096).astype(np.float16)
+    waypoint_idx = 32000 % 599  # Simulate waypoint position within sequence
+
+    # Extract hidden state at waypoint position
+    ego_feature = full_hs[0, waypoint_idx, :]
+
+    assert ego_feature.shape == (4096,), f"Expected (4096,), got {ego_feature.shape}"
+    assert np.allclose(ego_feature, full_hs[0, waypoint_idx, :]), "Extraction mismatch"
```
Comment (brnguyen2) [inline examples/vla_hidden_states_export/tests/test_hidden_states.py]: `assert np.allclose(ego_feature, full_hs[0, waypoint_idx, :])` compares the slice to itself — it can never fail, so this test verifies nothing beyond the shape assert on the line above. Relatedly, `waypoint_idx = 32000 % 599` on line 62 is arbitrary; if the intent is "a caller-supplied token index", just pick one and say so.

File: examples/vla_hidden_states_export/export_hidden_states.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+import argparse
+import os
+import subprocess
+import sys
+
+
+def build_engine_v0x(model_dir, output_dir, dtype="fp16"):
+    """Build engine for v0.x TRT backend (requires mark_output patch)."""
+    cmd = [
+        "trtllm-build",
```
Comment (brnguyen2) [inline examples/vla_hidden_states_export/export_hidden_states.py]: `trtllm-build` no longer exists on `main` — `tensorrt_llm/commands/` has only `bench`, `eval`, and `serve`. This script can't run against the branch it's being added to. Please either drop it or move the whole v0.x flow into a clearly-labelled historical note in the README.

File: examples/vla_hidden_states_export/export_hidden_states.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
+"""
```
Comment (brnguyen2) [inline examples/vla_hidden_states_export/export_hidden_states.py:3]: New files need the NVIDIA copyright header (`AGENTS.md`, "CRITICAL" section). Applies to `inference_python.py` and `tests/test_hidden_states.py` too.

**Comment by hous-lab** [comment]
@brnguyen2 I've addressed both issues from your review:

1. **Branch no longer reverts merged work** - the diff is now +383 additions, 0 deletions, based on latest main
2. **All code quality issues fixed** - real diff hunks, ValueError instead of assert, type annotations, copyright headers, remove_input_padding shape note

Also, nvpohanh has approved this PR. Could you re-review when you have a moment? Happy to adjust if there's a preferred integration approach.

Thanks!
